File: LCX/ValidEndPoints/dsl/examples/authenticated_trading.js
const axios = require('axios');
const crypto = require('crypto');
const os = require('os');


async function get_account_info() {
  // GET /api/account
  const url = 'https://exchange-api.lcx.com/api/account';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/account';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_balances() {
  // GET /api/balances
  const url = 'https://exchange-api.lcx.com/api/balances';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/balances';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_asset_balance(asset) {
  // GET /api/balance
  const url = 'https://exchange-api.lcx.com/api/balance';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/balance';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_trading_fees() {
  // GET /api/fees
  const url = 'https://exchange-api.lcx.com/api/fees';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/fees';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_open_orders(offset, pair) {
  // GET /api/open
  const url = 'https://exchange-api.lcx.com/api/open';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/open';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_order_details(orderId) {
  // GET /api/order
  const url = 'https://exchange-api.lcx.com/api/order';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/order';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_order_history(offset, pair, fromDate, toDate, side, orderStatus, orderType) {
  // GET /api/orderHistory
  const url = 'https://exchange-api.lcx.com/api/orderHistory';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/orderHistory';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_user_trade_history(offset, pair, fromDate, toDate) {
  // GET /api/uHistory
  const url = 'https://exchange-api.lcx.com/api/uHistory';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/uHistory';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function place_order(Pair, Amount, Price, OrderType, Side, ClientOrderId) {
  // POST /api/create
  const url = 'https://exchange-api.lcx.com/api/create';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'POST' + '/api/create';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.post(url, {}, {headers});
    return response.data;

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

# PUT method not supported for javascript
# DELETE method not supported for javascript
# DELETE method not supported for javascript
async function check_balance_workflow() {
  // GET /api/balances
  const url = 'https://exchange-api.lcx.com/api/balances';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/balances';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

# Workflow: CheckBalanceWorkflow (sequence)

# None method not supported for javascript
async function place_order_workflow(BTC/USDC) {
  // GET /api/ticker
  const url = 'https://exchange-api.lcx.com/api/ticker';
  const headers = {'Content-Type': 'application/json'};

  try {
    const response = await axios.get(url, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

# Workflow: PlaceOrderWorkflow (sequence)

async function place_limit_order(BTC/USDC, 0.01, 40000, LIMIT, BUY) {
  // POST /api/create
  const url = 'https://exchange-api.lcx.com/api/create';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'POST' + '/api/create';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.post(url, {}, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_order_info(order_response) {
  // GET /api/order
  const url = 'https://exchange-api.lcx.com/api/order';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/order';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

# None method not supported for javascript
async function manage_open_orders_workflow(0) {
  // GET /api/orderHistory
  const url = 'https://exchange-api.lcx.com/api/orderHistory';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/orderHistory';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

# Workflow: ManageOpenOrdersWorkflow (sequence)

async function get_trade_history(0) {
  // GET /api/uHistory
  const url = 'https://exchange-api.lcx.com/api/uHistory';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/uHistory';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

# None method not supported for javascript
# PUT method not supported for javascript
# DELETE method not supported for javascript
# None method not supported for javascript
async function account_monitoring_workflow() {
  // GET /api/balances
  const url = 'https://exchange-api.lcx.com/api/balances';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/balances';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

# Workflow: AccountMonitoringWorkflow (sequence)

async function get_open_orders(0) {
  // GET /api/open
  const url = 'https://exchange-api.lcx.com/api/open';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/open';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

async function get_fees() {
  // GET /api/fees
  const url = 'https://exchange-api.lcx.com/api/fees';
  const headers = {'Content-Type': 'application/json'};
  // HMAC-SHA256 Authentication
  const apiKey = process.env.LCX_API_KEY;
  const apiSecret = process.env.LCX_API_SECRET;
  const timestamp = Date.now().toString();
  const message = timestamp + 'GET' + '/api/fees';
  const signature = crypto
    .createHmac('sha256', apiSecret)
    .update(message)
    .digest('base64');
  headers['x-access-key'] = apiKey;
  headers['x-access-sign'] = signature;
  headers['x-access-timestamp'] = timestamp;

  try {
    const response = await axios.get(url, {headers});

  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

# None method not supported for javascript

File: LCX/ValidEndPoints/dsl/examples/authenticated_trading.cs
using System;
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json;
using System.Security.Cryptography;
using System.Text;


public static async Task<dynamic> get_account_info()
{
    // GET /api/account
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/account";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/account";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject(content);

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

public static async Task<dynamic> get_balances()
{
    // GET /api/balances
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/balances";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/balances";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject(content);

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

public static async Task<dynamic> get_asset_balance(asset)
{
    // GET /api/balance
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/balance";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/balance";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject(content);

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

public static async Task<dynamic> get_trading_fees()
{
    // GET /api/fees
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/fees";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/fees";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject(content);

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

public static async Task<dynamic> get_open_orders(offset, pair)
{
    // GET /api/open
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/open";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/open";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject(content);

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

public static async Task<dynamic> get_order_details(orderId)
{
    // GET /api/order
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/order";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/order";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject(content);

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

public static async Task<dynamic> get_order_history(offset, pair, fromDate, toDate, side, orderStatus, orderType)
{
    // GET /api/orderHistory
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/orderHistory";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/orderHistory";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject(content);

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

public static async Task<dynamic> get_user_trade_history(offset, pair, fromDate, toDate)
{
    // GET /api/uHistory
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/uHistory";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/uHistory";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject(content);

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

# POST method not supported for csharp
# PUT method not supported for csharp
# DELETE method not supported for csharp
# DELETE method not supported for csharp
public static async Task<dynamic> check_balance_workflow()
{
    // GET /api/balances
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/balances";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/balances";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

# Workflow: CheckBalanceWorkflow (sequence)

# None method not supported for csharp
public static async Task<dynamic> place_order_workflow(BTC/USDC)
{
    // GET /api/ticker
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/ticker";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

# Workflow: PlaceOrderWorkflow (sequence)

# POST method not supported for csharp
public static async Task<dynamic> get_order_info(order_response)
{
    // GET /api/order
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/order";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/order";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

# None method not supported for csharp
public static async Task<dynamic> manage_open_orders_workflow(0)
{
    // GET /api/orderHistory
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/orderHistory";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/orderHistory";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

# Workflow: ManageOpenOrdersWorkflow (sequence)

public static async Task<dynamic> get_trade_history(0)
{
    // GET /api/uHistory
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/uHistory";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/uHistory";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

# None method not supported for csharp
# PUT method not supported for csharp
# DELETE method not supported for csharp
# None method not supported for csharp
public static async Task<dynamic> account_monitoring_workflow()
{
    // GET /api/balances
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/balances";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/balances";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

# Workflow: AccountMonitoringWorkflow (sequence)

public static async Task<dynamic> get_open_orders(0)
{
    // GET /api/open
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/open";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/open";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

public static async Task<dynamic> get_fees()
{
    // GET /api/fees
    using (var client = new HttpClient())
    {
        var url = "https://exchange-api.lcx.com/api/fees";
        client.DefaultRequestHeaders.Add("Content-Type", "application/json");
        var apiKey = Environment.GetEnvironmentVariable("LCX_API_KEY");
        var apiSecret = Environment.GetEnvironmentVariable("LCX_API_SECRET");
        var timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        var message = timestamp + "GET" + "/api/fees";
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(apiSecret)))
        {
            var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            client.DefaultRequestHeaders.Add("x-access-key", apiKey);
            client.DefaultRequestHeaders.Add("x-access-sign", signature);
            client.DefaultRequestHeaders.Add("x-access-timestamp", timestamp);
        }

        try
        {
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();

        }
        catch (HttpRequestException e)
        {
            Console.WriteLine("Error: " + e.Message);
            throw;
        }
    }
}

# None method not supported for csharp
